gnovm/pkg/doc/dirs.go

```diff
@@ -30,15 +34,82 @@ type bfsDirs struct {
 }
 
 // newDirs begins scanning the given stdlibs directory.
-func newDirs(dirs ...string) *bfsDirs {
+// dirs are "gopath-like" directories, such as @/gnovm/stdlibs and @/examples.
+// modDirs are user directories, expected to have gno.mod files
+func newDirs(dirs []string, modDirs []string) *bfsDirs {
 	d := &bfsDirs{
 		hist: make([]bfsDir, 0, 256),
 		scan: make(chan bfsDir),
 	}
-	go d.walk(dirs)
+
+	roots := make([]bfsDir, 0, len(dirs)+len(modDirs))
+	for _, dir := range dirs {
+		roots = append(roots, bfsDir{
+			dir:        dir,
+			importPath: "",
+		})
+	}
+
+	for _, mdir := range modDirs {
+		gm, err := tryParseGnoMod(filepath.Join(mdir, "gno.mod"))
+		if err != nil {
+			log.Printf("%v", err)
+			continue
+		}
+		roots = append(roots, bfsDir{
+			dir:        mdir,
+			importPath: gm.Module.Mod.Path,
+		})
+		roots = append(roots, getGnoModDirs(gm)...)
+	}
+
+	go d.walk(roots)
 	return d
 }
 
+// tries to parse gno mod file.
+// second return parameter is whether gno.mod exists.
+func tryParseGnoMod(fname string) (*gnomod.File, error) {
+	file, err := os.Stat(fname)
+	// early exit for errors and non-file fname
+	if err != nil || file.IsDir() {
+		if err == nil {
+			return nil, fmt.Errorf("invalid gno.mod at %q: is a directory", fname)
+		}
+		// no need for filename, os errors contain that already
+		return nil, fmt.Errorf("could not read go.mod file: %w", err)
+	}
```

Comment:
Just a nitpick; looks simpler and have fewer nested conditions; `s/read go.mod/read gno.mod`

```suggestion
	file, err := os.Stat(fname)
	if err != nil {
		return nil, fmt.Errorf("could not read gno.mod file: %w", err)
	}
	if file.IsDir() {
		return nil, fmt.Errorf("invalid gno.mod at %q: is a directory", fname)
	}
```
---

I personally don't like the idea of checking a condition and then checking its opposite within the inner condition. For instance:
```go
if err != nil || ... {
    if err == nil {
        ...
    }
    ...
}
```